enhancements/cloud-integration/infrastructure-external-platform-type.md

```diff
@@ -0,0 +1,560 @@
+the actual handling of such objects happens in a provider-specific [cloud controller manager](https://kubernetes.io/docs/concepts/architecture/cloud-controller/).
+
+#### Cluster Network Operator
+
+Speaking about Cluster Network Operator, several things depends on PlatformType there.
+
+- Component called CNCC ([cloud-network-config-controller](https://github.com/openshift/cloud-network-config-controller))
+  contains the majority of platform-specific logic. CNO makes the decision about CNCC deployment based on
+  the PlatformType value. Mainly CNCC itself is responsible for attaching/detaching private IP addresses to VMs associated with
+  Kubernetes nodes, which require interaction with cloud-provider APIs.
+  Currently, CNCC [deploys](https://github.com/openshift/cluster-network-operator/blob/33287d5d9580556ca1757fd4f20b1a6b99e68786/pkg/network/cloud_network.go#L29)
+  on GCP, Azure, AWS, and OpenStack platforms, other platforms, such as IBM or Alibaba do not engage CNCC for the moment.
+
+- There are also several platform specific hacks,
+  like the [access restriction to a metadata service](https://github.com/openshift/ovn-kubernetes/blob/8d5db0bae387de423eb30db75142b0ad87df4f3f/go-controller/pkg/cni/OCP_HACKS.go#L18),
+  but it is not entirely connected with the operator itself and is more CNI plugin specific.
+
+For the [phase 1](#implementation-phases) of this project, there seems to be no particular action or API knobs needed regarding
+the addition of the "External" platform type. Just ensure that CNO is non-reactive to the "External" platform type
+and behaves the same as in the "None" platform case.
+In the future, we possibly might want to make CNO more tolerant of partner CNCC implementations
+and design a way for configuring platform-specific CNI behaviour.
+
+### Workflow Description
+
+During [phase 1](#implementation-phases) proposed changes intended to have no effect and the "External"
+platform type should be handled the same as the "None", so no specific user interaction is expected.
+
+### API Extensions
+
+A new optional constant of `PlatformType` type will be added to the "openshift/api".
+
+```go
+const (
+    ...
+    // ExternalPlatformType represent generic infrastructure provider. Provider-specific components should be supplemented separately.
+    ExternalPlatformType PlatformType = "External"
+    ...
+)
+```
+
+Additionally, the respective external platform spec and status should be added to the infrastructure resource.
+
+```go
+type CloudControllerManagerMode string
+
+const (
+    // Cloud Controller Manager is enabled and expected to be supplied.
+    // Signaling MCO to set `--cloud-provider=external` flag to the kubelets.
+    CloudControllerManagerExternal CloudControllerManagerState = "External"
+    // No Cloud Controller Manager is expected to be supplied.
+    // Signaling MCO not to set `--cloud-provider` flag to the kubelets.
+    CloudControllerManagerNone CloudControllerManagerState = "None"
```

Comment:
This (vs. the `None` platform) is imagining a world where there is no external cloud-provider for the kubelet and other folks who use `--cloud-provider`, but there is an external cloud provider for... I'm not actually sure which component that would be.  Maybe we can skip `CloudControllerManagerNone` until it has a clearer use-case, because adding new enum entries isn't a breaking change?

Reply:
> but there is an external cloud provider for...

MachineAPI, for example. Currently we have this situation with the nutanix, MAPI provider was landed, but ccm is still in progress and we operating without it.

Reply:
Now that this thread is hung on an obsolete line (with the `State` -> `Mode` change), let's move discussion [over here][1] and mark this thread resolved.

[1]: https://github.com/openshift/enhancements/pull/1234#discussion_r1002902884